bpo-29779: new environment variable PYTHONHISTORY.

0xl3vi opened the pull request:
if this variable is set the user can change the location of
a ~/.python_history file without adding hook to PYTHONSTARTUP.

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA. This is necessary for legal reasons before we can look at your contribution. Please follow these steps to help rectify the issue:

1. If you don't have an account on b.p.o, please [create one](http://bugs.python.org/user?@template=register)
2. Make sure your GitHub username is listed in ["Your Details" at b.p.o](https://cloud.githubusercontent.com/assets/2680980/23276970/d14a380c-f9d1-11e6-883d-e13b6b211239.png)
3. If you have not already done so, please sign the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/). The "bugs.python.org username " requested by the form is the "Login name" field under "Your Details".
4. If you just signed the CLA, please **wait at least one US business day** and then check "Your Details" on [bugs.python.org](https://bugs.python.org/) to see if your account has been marked as having signed the CLA (the delay is due to a person having to manually check your signed CLA)
5. Reply here saying you have completed the above steps


Thanks again to your contribution and we look forward to looking at it!

ghost commented:
I signed the CLA :wink:

warsaw commented:
This is a nice change, and it would probably let me get rid of my $PYTHONSTARTUP setting.  I wonder if `python3 -h` should mention this new envar?

ghost commented:
@warsaw I added this to Misc/python.man

warsaw commented:
@0xl3vi Thanks, that looks good.  It's still not mentioned in `python3 -h` but I'm not entirely sure it needs to.  There are other `PYTHON*` environment variables not mentioned in the `--help` output.

warsaw commented on Lib/site.py:
```diff
@@ -412,14 +424,13 @@ def register_readline():
-            history = os.path.join(os.path.expanduser('~'),
-                                   '.python_history')
+            history = set_history_file()
+            
```
Minor style nit: You don't need this extra blank line.

warsaw reviewed (approved):
LGTM, with one very minor style nit.

ghost commented:
@warsaw I updated main.c for `python -h`,
blank line fixed.

warsaw commented:
The changes look great to me, but I'm going to wait just a bit before merging this.

* I'd like to see if any other core dev has an opinion about the change
* The AppVeyor test is failing, although I suspect that is unrelated to your change

I don't have a Windows environment to test that on.  Maybe @zooba can weigh in on that.  Alternatively, try rebasing against master if it's been fixed there.

If there are no other follow ups by this weekend, I'll merge your change.

Can you please also add a `Misc/NEWS` entry (knowing that our workflow for that kind of sucks and we might get conflicts)?

ned-deily reviewed (changes requested):
Environment variables are also documented in the Python Setup and Usage document (https://docs.python.org/3/using/cmdline.html#environment-variables) which is generated from Doc/using/cmdline.rst.

ned-deily commented:
Also there should be an issue opened on bugs.python.org to document this and the issue number added to the title of this pull request.

ghost commented:
@warsaw I tried rebasing against master and the build passed!
and Misc/NEWS updated.

@ned-deily I updated cmdline.rst and created new issue http://bugs.python.org/issue29779

P.S: I need to add bpo-XXX or just the issue number in bugs.python.org?

warsaw commented on Lib/site.py:
```diff
@@ -372,6 +372,18 @@ def setcopyright():
+    if history_file:
+        return history_file
+    return os.path.join(os.path.expanduser('~'),
+        '.python_history')
```
You could also do something like:

```python
    return os.environ.get(
        'PYTHONHISTORY',
        os.path.expanduser(os.path.join('~', '.python_history')))
```

Also `set_history_file()` is a little bit of a misnomer since it's not actually setting it, but calculating it instead.  OTOH, I bet the more compact form could just be inlined below.

warsaw commented on Misc/NEWS:
```diff
@@ -10,6 +10,9 @@ What's New in Python 3.7.0 alpha 1?
 Core and Builtins
 -----------------
 
+- Issue #29779: New environment variable PYTHONHISTORY if
```
Change `Issue #29779` to `bpo-29779`.  That's the new style of marking issue numbers.

warsaw reviewed (changes requested):
Just a couple of other thoughts on the implementation and NEWS file entry.

warsaw commented:
Oh, and it would be nice to add a test for codecov.

ned-deily commented:
Thanks for addressing my comments!

ghost commented:
@warsaw all done, I changed set_history_file() to gethistoryfile().
and added test.

warsaw commented:
@0xl3vi We still need a test.

warsaw commented:
Oh and yay for Misc/NEWS conflicts. :(

warsaw commented:
Also note [this report](http://bugs.python.org/issue29779#msg289497) which I can confirm.

ghost commented:
@warsaw sorry for the dealy.
what tests we need add?

regarding [this](http://bugs.python.org/issue29779#msg289497)
If the PYTHONHISTORY is empty it will fall back to .python_history.
there is no error on my build

warsaw commented:
@0xl3vi At the very least, I think a test covering `gethistoryfile()` would be useful.   If you can mock out enough of the readline stuff (or other safeguards so as not to interfere with the environment of a person running the tests), then adding some coverage for `enablerlcompleter()` would be nice.

The bug happens when the file pointed to by `PYTHONHISTORY` does not exist.

Commit:
bpo-29779: new environment variable PYTHONHISTORY.

this patch adds new environment variable PYTHONHISTORY,
with this you can change the location of a python_history file,
without using readline hook in PYTHONSTARTUP.

In case PYTHONHISTORY will be empty or not set,
it wil fall back to the default history file.

ghost commented:
@warsaw OK updated.

* In case PYTHONHISTORY will be empty or not set,
it wil fall back to the default history file.
if the file pointed by PYTHONHISTORY does not exist readline will create one.

The test checks if gethistoryfile() reads the variable.
Bash will expand '~' so we dont need to to test the path.

what do you think?

berkerpeksag commented on Doc/using/cmdline.rst:
```diff
@@ -713,6 +713,13 @@ conflict.
+.. envvar:: PYTHONHISTORY
+
+   If set to a non-empty string, you can change the location of a python_history
+   file, by default it will be in ~/.python_history.
```
\`\`~/.python_history\`\`

berkerpeksag commented on Doc/using/cmdline.rst:
```diff
@@ -713,6 +713,13 @@ conflict.
 
+.. envvar:: PYTHONHISTORY
+
+   If set to a non-empty string, you can change the location of a python_history
```
"a \`\`.python_history\`\` file" or just "a history file".

berkerpeksag commented on Lib/site.py:
```diff
@@ -372,6 +372,17 @@ def setcopyright():
+     if not use the default ~/.python_history file.
+     """
+    h = os.environ.get("PYTHONHISTORY")
+    if h != '':
```
What if `os.environ.get("PYTHONHISTORY")` returns `None`?

berkerpeksag commented on Lib/test/test_site.py:
```diff
@@ -259,6 +259,10 @@ def test_getsitepackages(self):
             self.assertEqual(dirs[1], wanted)
 
+    def test_gethistoryfile(self):
+        os.environ['PYTHONHISTORY'] = "xoxo"
```
Please use `EnvironmentVarGuard` in `test.support` to set env variables.

berkerpeksag commented on Misc/NEWS:
```diff
@@ -10,6 +10,9 @@ What's New in Python 3.7.0 alpha 1?
 -----------------
 
+- bpo-29779: New environment variable PYTHONHISTORY if
+  this is set you can change the location of a python_history file.
```
Please add "Patch by Your Name."

berkerpeksag commented on Misc/python.man:
```diff
@@ -431,6 +431,10 @@ values.
 the value 0 will disable hash randomization.
+.IP PYTHONHISTORY
+If this is set you can change the location of a
+python_history file, by default it will be ~/.python_history.
```
File names should be italic in man pages so `\fI~/.python_history\fI` or

```man
python_history file, by default it will be
.IR ~/.python_history .
```

You may need to escape ~ or / characters.

berkerpeksag commented on Modules/main.c:
```diff
@@ -105,7 +105,8 @@ static const char usage_6[] =
 "   on Python memory allocators. Use PYTHONMALLOC=debug to install debug\n"
-"   hooks.\n";
+"   hooks.\n"
+"PYTHONHISTORY: If this is set, you can change the location of a python_history file.\n";
```
This line is a bit long.

berkerpeksag reviewed (changes requested):
This also needs to be documented in `Doc/whatsnew/3.7.rst` (and please add "(Contributed by Your Name.)")

Thanks!

ghost commented:
OK, I don't have time to do it.
If someone else want to implement this please do so.

warsaw commented:
@0xl3vi Hi.  I'm sorry this has taken so long.  Are you saying in your comment that you don't have time to finish this branch?  If so, that's okay and we greatly appreciate your contribution so far.  I'm willing to finish this branch for you if you'd like.

ghost commented:
@warsaw sure if you want.
thanks :+1:
